File: src/schemas/pattern/display.ts
import { field, fixedArray, option, vec } from "@dao-xyz/borsh";

export class Symbols {
  @field({ type: option("u16") })
  full?: number;

  @field({ type: option("u16") })
  petite?: number;

  @field({ type: option("u16") })
  half?: number;

  @field({ type: option("u16") })
  quarter?: number;

  @field({ type: option("u16") })
  frenchKnot?: number;

  @field({ type: option("u16") })
  bead?: number;

  constructor(data: Symbols) {
    this.full = data.full;
    this.petite = data.petite;
    this.half = data.half;
    this.quarter = data.quarter;
    this.frenchKnot = data.frenchKnot;
    this.bead = data.bead;
  }
}

export class SymbolSettings {
  @field({ type: fixedArray("u16", 2) })
  screenSpacing: [number, number];

  @field({ type: fixedArray("u16", 2) })
  printerSpacing: [number, number];

  @field({ type: "bool" })
  scaleUsingMaximumFontWidth: boolean;

  @field({ type: "bool" })
  scaleUsingFontHeight: boolean;

  @field({ type: "u16" })
  stitchSize: number;

  @field({ type: "u16" })
  smallStitchSize: number;

  @field({ type: "bool" })
  drawSymbolsOverBackstitches: boolean;

  @field({ type: "bool" })
  showStitchColor: boolean;

  @field({ type: "bool" })
  useLargeHalfStitchSymbol: boolean;

  @field({ type: "bool" })
  useTrianglesBehindQuarterStitches: boolean;

  constructor(data: SymbolSettings) {
    this.screenSpacing = data.screenSpacing;
    this.printerSpacing = data.printerSpacing;
    this.scaleUsingMaximumFontWidth = data.scaleUsingMaximumFontWidth;
    this.scaleUsingFontHeight = data.scaleUsingFontHeight;
    this.stitchSize = data.stitchSize;
    this.smallStitchSize = data.smallStitchSize;
    this.drawSymbolsOverBackstitches = data.drawSymbolsOverBackstitches;
    this.showStitchColor = data.showStitchColor;
    this.useLargeHalfStitchSymbol = data.useLargeHalfStitchSymbol;
    this.useTrianglesBehindQuarterStitches = data.useTrianglesBehindQuarterStitches;
  }
}

export class SymbolFormat {
  @field({ type: "bool" })
  useAltBgColor: boolean;

  @field({ type: "string" })
  bgColor: string;

  @field({ type: "string" })
  fgColor: string;

  constructor(data: SymbolFormat) {
    this.useAltBgColor = data.useAltBgColor;
    this.bgColor = data.bgColor;
    this.fgColor = data.fgColor;
  }
}

export class LineStitchFormat {
  @field({ type: "bool" })
  useAltColor: boolean;

  @field({ type: "string" })
  color: string;

  @field({ type: "u8" })
  style: LineStitchStyle;

  @field({ type: "f32" })
  thickness: number;

  constructor(data: LineStitchFormat) {
    this.useAltColor = data.useAltColor;
    this.color = data.color;
    this.style = data.style;
    this.thickness = data.thickness;
  }
}

export class NodeStitchFormat {
  @field({ type: "bool" })
  useDotStyle: boolean;

  @field({ type: "bool" })
  useAltColor: boolean;

  @field({ type: "string" })
  color: string;

  @field({ type: "f32" })
  diameter: number;

  constructor(data: NodeStitchFormat) {
    this.useDotStyle = data.useDotStyle;
    this.useAltColor = data.useAltColor;
    this.color = data.color;
    this.diameter = data.diameter;
  }
}

export class FontFormat {
  @field({ type: option("string") })
  fontName?: string;

  @field({ type: "bool" })
  bold: boolean;

  @field({ type: "bool" })
  italic: boolean;

  @field({ type: "u16" })
  stitchSize: number;

  @field({ type: "u16" })
  smallStitchSize: number;

  constructor(data: FontFormat) {
    this.fontName = data.fontName;
    this.bold = data.bold;
    this.italic = data.italic;
    this.stitchSize = data.stitchSize;
    this.smallStitchSize = data.smallStitchSize;
  }
}

export class Formats {
  @field({ type: SymbolFormat })
  symbol: SymbolFormat;

  @field({ type: LineStitchFormat })
  backstitch: LineStitchFormat;

  @field({ type: LineStitchFormat })
  straightstitch: LineStitchFormat;

  @field({ type: NodeStitchFormat })
  frenchKnot: NodeStitchFormat;

  @field({ type: NodeStitchFormat })
  bead: NodeStitchFormat;

  @field({ type: LineStitchFormat })
  specialstitch: LineStitchFormat;

  @field({ type: FontFormat })
  font: FontFormat;

  constructor(data: Formats) {
    this.symbol = data.symbol;
    this.backstitch = data.backstitch;
    this.straightstitch = data.straightstitch;
    this.frenchKnot = data.frenchKnot;
    this.bead = data.bead;
    this.specialstitch = data.specialstitch;
    this.font = data.font;
  }
}

export class GridLineStyle {
  @field({ type: "string" })
  color: string;

  @field({ type: "f32" })
  thickness: number;

  constructor(data: GridLineStyle) {
    this.color = data.color;
    this.thickness = data.thickness;
  }
}

export class Grid {
  @field({ type: "u16" })
  majorLineEveryStitches: number;

  @field({ type: GridLineStyle })
  minorScreenLines: GridLineStyle;

  @field({ type: GridLineStyle })
  majorScreenLines: GridLineStyle;

  @field({ type: GridLineStyle })
  minorPrinterLines: GridLineStyle;

  @field({ type: GridLineStyle })
  majorPrinterLines: GridLineStyle;

  constructor(data: Grid) {
    this.majorLineEveryStitches = data.majorLineEveryStitches;
    this.minorScreenLines = data.minorScreenLines;
    this.majorScreenLines = data.majorScreenLines;
    this.minorPrinterLines = data.minorPrinterLines;
    this.majorPrinterLines = data.majorPrinterLines;
  }
}

export class StitchOutline {
  @field({ type: option("string") })
  color?: string;

  @field({ type: "u16" })
  colorPercentage: number;

  @field({ type: "f32" })
  thickness: number;

  constructor(data: StitchOutline) {
    this.color = data.color;
    this.colorPercentage = data.colorPercentage;
    this.thickness = data.thickness;
  }
}

export class DefaultStitchStrands {
  @field({ type: "u16" })
  full: number;

  @field({ type: "u16" })
  petite: number;

  @field({ type: "u16" })
  half: number;

  @field({ type: "u16" })
  quarter: number;

  @field({ type: "u16" })
  back: number;

  @field({ type: "u16" })
  straight: number;

  @field({ type: "u16" })
  frenchKnot: number;

  @field({ type: "u16" })
  special: number;

  constructor(data: DefaultStitchStrands) {
    this.full = data.full;
    this.petite = data.petite;
    this.half = data.half;
    this.quarter = data.quarter;
    this.back = data.back;
    this.straight = data.straight;
    this.frenchKnot = data.frenchKnot;
    this.special = data.special;
  }
}

export class StitchSettings {
  @field({ type: DefaultStitchStrands })
  defaultStrands: DefaultStitchStrands;

  @field({ type: fixedArray("f32", 13) })
  displayThickness: number[];

  constructor(data: StitchSettings) {
    this.defaultStrands = data.defaultStrands;
    this.displayThickness = data.displayThickness;
  }
}

export class DisplaySettings {
  @field({ type: "string" })
  defaultStitchFont: string;

  @field({ type: vec(Symbols) })
  symbols: Symbols[];

  @field({ type: SymbolSettings })
  symbolSettings: SymbolSettings;

  @field({ type: vec(Formats) })
  formats: Formats[];

  @field({ type: Grid })
  grid: Grid;

  @field({ type: "u8" })
  view: number;

  @field({ type: "u16" })
  zoom: number;

  @field({ type: "bool" })
  showGrid: boolean;

  @field({ type: "bool" })
  showRulers: boolean;

  @field({ type: "bool" })
  showCenteringMarks: boolean;

  @field({ type: "bool" })
  showFabricColorsWithSymbols: boolean;

  @field({ type: "bool" })
  gapsBetweenStitches: boolean;

  @field({ type: "bool" })
  outlinedStitches: boolean;

  @field({ type: StitchOutline })
  stitchOutline: StitchOutline;

  @field({ type: StitchSettings })
  stitchSettings: StitchSettings;

  constructor(data: DisplaySettings) {
    this.defaultStitchFont = data.defaultStitchFont;
    this.symbols = data.symbols;
    this.symbolSettings = data.symbolSettings;
    this.formats = data.formats;
    this.grid = data.grid;
    this.view = data.view;
    this.zoom = data.zoom;
    this.showGrid = data.showGrid;
    this.showRulers = data.showRulers;
    this.showCenteringMarks = data.showCenteringMarks;
    this.showFabricColorsWithSymbols = data.showFabricColorsWithSymbols;
    this.gapsBetweenStitches = data.gapsBetweenStitches;
    this.outlinedStitches = data.outlinedStitches;
    this.stitchOutline = data.stitchOutline;
    this.stitchSettings = data.stitchSettings;
  }
}

export const enum LineStitchStyle {
  Solid = 0,
  Barred = 1,
  Dotted = 2,
  ChainDotted = 3,
  Dashed = 4,
  Outlined = 5,
  Zebra = 6,
  ZigZag = 7,
  Morse = 8,
}

export const enum View {
  Stitches = 0,
  Symbols = 1,
  Solid = 2,
  Information = 3,
  MachineEmbInfo = 4,
}

File: src-tauri/src/pattern/display.rs
use borsh::{BorshDeserialize, BorshSerialize};
use ordered_float::NotNan;

pub type Points = NotNan<f32>;

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct DisplaySettings {
  pub default_stitch_font: String,
  pub symbols: Vec<Symbols>,
  pub symbol_settings: SymbolSettings,
  pub formats: Vec<Formats>,
  pub grid: Grid,
  pub view: View,
  pub zoom: u16,
  pub show_grid: bool,
  pub show_rulers: bool,
  pub show_centering_marks: bool,
  pub show_fabric_colors_with_symbols: bool,
  pub gaps_between_stitches: bool,
  pub outlined_stitches: bool,
  pub stitch_outline: StitchOutline,
  pub stitch_settings: StitchSettings,
}

impl DisplaySettings {
  pub fn new(palette_size: usize) -> Self {
    Self {
      default_stitch_font: String::from("CrossStitch3"),
      symbols: vec![Symbols::default(); palette_size],
      symbol_settings: SymbolSettings::default(),
      formats: vec![Formats::default(); palette_size],
      grid: Grid::default(),
      view: View::Solid,
      zoom: 100,
      show_grid: true,
      show_rulers: true,
      show_centering_marks: true,
      show_fabric_colors_with_symbols: false,
      gaps_between_stitches: false,
      outlined_stitches: true,
      stitch_outline: StitchOutline::default(),
      stitch_settings: StitchSettings::default(),
    }
  }
}

#[derive(Debug, Default, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct Symbols {
  pub full: Option<u16>,
  pub petite: Option<u16>,
  pub half: Option<u16>,
  pub quarter: Option<u16>,
  pub french_knot: Option<u16>,
  pub bead: Option<u16>,
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct SymbolSettings {
  pub screen_spacing: (u16, u16),
  pub printer_spacing: (u16, u16),
  pub scale_using_maximum_font_width: bool,
  pub scale_using_font_height: bool,
  pub stitch_size: u16,
  pub small_stitch_size: u16,
  pub draw_symbols_over_backstitches: bool,
  pub show_stitch_color: bool,
  pub use_large_half_stitch_symbol: bool,
  pub use_triangles_behind_quarter_stitches: bool,
}

impl Default for SymbolSettings {
  fn default() -> Self {
    Self {
      screen_spacing: (1, 1),
      printer_spacing: (1, 1),
      scale_using_maximum_font_width: true,
      scale_using_font_height: true,
      stitch_size: 100,
      small_stitch_size: 60,
      draw_symbols_over_backstitches: false,
      show_stitch_color: false,
      use_large_half_stitch_symbol: false,
      use_triangles_behind_quarter_stitches: false,
    }
  }
}

#[derive(Debug, Default, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct Formats {
  pub symbol: SymbolFormat,
  pub back: LineFormat,
  pub straight: LineFormat,
  pub french: NodeFormat,
  pub bead: NodeFormat,
  pub special: LineFormat,
  pub font: FontFormat,
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct SymbolFormat {
  pub use_alt_bg_color: bool,
  pub bg_color: String,
  pub fg_color: String,
}

impl Default for SymbolFormat {
  fn default() -> Self {
    Self {
      use_alt_bg_color: false,
      bg_color: String::from("FFFFFF"),
      fg_color: String::from("000000"),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct LineFormat {
  pub use_alt_color: bool,
  pub color: String,
  pub style: LineStyle,
  pub thickness: Points,
}

impl Default for LineFormat {
  fn default() -> Self {
    Self {
      use_alt_color: false,
      color: String::from("000000"),
      style: LineStyle::Solid,
      thickness: NotNan::new(1.0).unwrap(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
#[borsh(use_discriminant = true)]
pub enum LineStyle {
  Solid = 0,
  Barred = 1,
  Dotted = 2,
  ChainDotted = 3,
  Dashed = 4,
  Outlined = 5,
  Zebra = 6,
  ZigZag = 7,
  Morse = 8,
}

impl From<u16> for LineStyle {
  fn from(value: u16) -> Self {
    match value {
      // These are the values used by Pattern Maker.
      0 | 5 => LineStyle::Solid,
      1 | 7 => LineStyle::Barred,
      2 | 6 => LineStyle::Dotted,
      11 => LineStyle::ChainDotted,
      3 | 8 => LineStyle::Dashed,
      9 => LineStyle::Outlined,
      10 => LineStyle::Zebra,
      12 => LineStyle::ZigZag,
      4 => LineStyle::Morse,
      _ => panic!("Invalid LineStyle value: {value}"),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct NodeFormat {
  pub use_dot_style: bool,
  pub use_alt_color: bool,
  pub color: String,
  pub diameter: Points,
}

impl Default for NodeFormat {
  fn default() -> Self {
    Self {
      use_dot_style: true,
      use_alt_color: false,
      color: String::from("000000"),
      diameter: NotNan::new(1.0).unwrap(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct FontFormat {
  pub font_name: Option<String>,
  pub bold: bool,
  pub italic: bool,
  pub stitch_size: u16,
  pub small_stitch_size: u16,
}

impl Default for FontFormat {
  fn default() -> Self {
    Self {
      font_name: None,
      bold: false,
      italic: false,
      stitch_size: 100,
      small_stitch_size: 60,
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct Grid {
  pub major_line_every_stitches: u16,
  pub minor_screen_lines: GridLineStyle,
  pub major_screen_lines: GridLineStyle,
  pub minor_printer_lines: GridLineStyle,
  pub major_printer_lines: GridLineStyle,
}

impl Default for Grid {
  fn default() -> Self {
    Self {
      major_line_every_stitches: 10,
      minor_screen_lines: GridLineStyle {
        color: String::from("000000"),
        thickness: NotNan::new(0.072).unwrap(),
      },
      major_screen_lines: GridLineStyle {
        color: String::from("000000"),
        thickness: NotNan::new(0.072).unwrap(),
      },
      minor_printer_lines: GridLineStyle {
        color: String::from("000000"),
        thickness: NotNan::new(0.144).unwrap(),
      },
      major_printer_lines: GridLineStyle {
        color: String::from("000000"),
        thickness: NotNan::new(0.504).unwrap(),
      },
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct GridLineStyle {
  pub color: String,
  pub thickness: Points,
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
#[borsh(use_discriminant = true)]
pub enum View {
  Stitches = 0,
  Symbols = 1,
  Solid = 2,
  Information = 3,
  MachineEmbInfo = 4,
}

impl From<u16> for View {
  fn from(value: u16) -> Self {
    match value {
      0 => View::Stitches,
      1 => View::Symbols,
      2 => View::Solid,
      3 => View::Information,
      5 => View::MachineEmbInfo,
      _ => panic!("Invalid View value: {value}"),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct StitchOutline {
  pub color: Option<String>,
  pub color_percentage: u16,
  pub thickness: Points,
}

impl Default for StitchOutline {
  fn default() -> Self {
    Self {
      color: None,
      color_percentage: 80,
      thickness: NotNan::new(0.2).unwrap(),
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct StitchSettings {
  pub default_strands: DefaultStitchStrands,
  /// 1..=12 - strands, 13 - french knot.
  pub display_thickness: [Points; 13],
}

impl Default for StitchSettings {
  fn default() -> Self {
    Self {
      default_strands: DefaultStitchStrands::default(),
      display_thickness: [
        NotNan::new(1.0).unwrap(), // 1 strand
        NotNan::new(1.5).unwrap(), // 2 strands
        NotNan::new(2.5).unwrap(), // 3 strands
        NotNan::new(3.0).unwrap(), // 4 strands
        NotNan::new(3.5).unwrap(), // 5 strands
        NotNan::new(4.0).unwrap(), // 6 strands
        NotNan::new(4.5).unwrap(), // 7 strands
        NotNan::new(5.0).unwrap(), // 8 strands
        NotNan::new(5.5).unwrap(), // 9 strands
        NotNan::new(6.0).unwrap(), // 10 strands
        NotNan::new(6.5).unwrap(), // 11 strands
        NotNan::new(7.0).unwrap(), // 12 strands
        NotNan::new(4.0).unwrap(), // French knot
      ],
    }
  }
}

#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
pub struct DefaultStitchStrands {
  pub full: u16,
  pub petite: u16,
  pub half: u16,
  pub quarter: u16,
  pub back: u16,
  pub straight: u16,
  pub special: u16,
}

impl Default for DefaultStitchStrands {
  fn default() -> Self {
    Self {
      full: 2,
      petite: 2,
      half: 2,
      quarter: 2,
      back: 1,
      straight: 1,
      special: 2,
    }
  }
}
